include/private/soci-compiler.h

```diff
@@ -36,4 +36,22 @@
 #   define SOCI_GCC_WARNING_RESTORE(x)
 #endif
 
+// SOCI_FALLTHROUGH macro can be used on both g++, mingw, and msvc to generate the
```

Comment:
Sorry, the conditions here don't seem right in several places. Compare with

https://github.com/wxWidgets/wxWidgets/blob/89b38ded388302fd6842f36405dfab27a6c34e9c/include/wx/defs.h#L258-L268

which, I believe, is correct.

Reply:
I tried to bring over the wxWidgets macros. Just SOCI'fying the macro names like this:

```
#if defined(_MSVC_LANG)
/*
   We want to always use the really supported C++ standard when using MSVC
   recent enough to define _MSVC_LANG, even if /Zc:__cplusplus option is not
   used, but unfortunately we can't just redefine __cplusplus as _MSVC_LANG
   because this is not allowed by the standard and, worse, doesn't work in
   practice (it results in a warning and nothing else).

   So, instead, we define a macro for testing __cplusplus which also works in
   this case.
*/
    #define SOCI_CHECK_CXX_STD(ver) (_MSVC_LANG >= (ver))
#elif defined(__cplusplus)
    #define SOCI_CHECK_CXX_STD(ver) (__cplusplus >= (ver))
#else
    #define SOCI_CHECK_CXX_STD(ver) 0
#endif


#ifndef SOCI_HAS_CLANG_FEATURE
#   ifndef __has_feature
#       define SOCI_HAS_CLANG_FEATURE(x) 0
#   else
#       define SOCI_HAS_CLANG_FEATURE(x) __has_feature(x)
#   endif
#endif

#pragma message(__cplusplus)
#if SOCI_CHECK_CXX_STD(201703L)
#    define SOCI_FALLTHROUGH [[fallthrough]]
#elif defined(__has_warning) && SOCI_HAS_CLANG_FEATURE(cxx_attributes)
#    define SOCI_FALLTHROUGH [[clang::fallthrough]]
#endif

#ifndef SOCI_FALLTHROUGH
    #define SOCI_FALLTHROUGH ((void)0)
#endif
```

However, when I compile it on my platform I get...

```
[ 82%] Building CXX object src/backends/postgresql/CMakeFiles/soci_postgresql_static.dir/session.cpp.o
In file included from /root/soci/src/backends/postgresql/session.cpp:12:
/root/soci/include/private/soci-compiler.h:66:17: warning: expected a string after ‘#pragma message’ [-Wpragmas]
   66 | #pragma message(VAR_NAME_VALUE(__cplusplus))
      |                 ^~~~~~~~~~~~~~
/root/soci/src/backends/postgresql/session.cpp: In function ‘std::vector<std::__cxx11::basic_string<char> > {anonymous}::get_schema_names(PGconn*)’:
/root/soci/src/backends/postgresql/session.cpp:72:21: warning: this statement may fall through [-Wimplicit-fallthrough=]
   72 |                     if (!quoted)
      |                     ^~
/root/soci/src/backends/postgresql/session.cpp:82:17: note: here
   82 |                 default:
      |                 ^~~~~~~
[ 83%] Linking CXX static library ../../../lib/libsoci_postgresql.a
```

This means that I'm not getting `[[fallthrough]]` declared.
Previously I tried to use the BOOST_FALLTHROUGH macro which gave me the same result. So, I'm guessing that there's something funny going on with my platform. Not sure what though.

Are we aiming at using these macro definitions (with a change of the comments) - or should I try to dig further into this?

Reply:
I just tried building from scratch in a ubuntu docker container. It does the same thing - so not just me.
It must be something with G++. Any ideas?

Reply:
What is your compiler and which compilation options (notable `-std=xxx`) do you use?

Reply:
I'm using g++ on ubuntu. 
`g++ (Ubuntu 13.2.0-23ubuntu4) 13.2.0`

Using default soci install:
```
mkdir soci/build && \
cd soci/build && \
cmake -G "Unix Makefiles" ../ --install-prefix=/usr && \
make
```

So nothing out of the ordinary...

Reply:
g++13 uses C++17 by default, so `SOCI_FALLTHROUGH` should be defined as `[[fallthrough]]` for it, please check — by inserting `#error` in the branches that should be taken — if all else fails, that this does happen in the header.

Reply:
g++ may support `[[fallthrough]]` in C++14 mode as an extension, I guess, but the real issue is that C++14 shouldn't be used in the first place.

There is

```cmake
if(NOT DEFINED CMAKE_CXX_STANDARD OR CMAKE_CXX_STANDARD LESS 14)
  set(CMAKE_CXX_STANDARD 14)
endif()
```

in SOCI CMakefile, which means that C++14 is used if `CMAKE_CXX_STANDARD` is not set at all. What we need to do is to only set the standard to 14 if the compiler doesn't support it in its default configuration. Unfortunately I don't think CMake allows to do this in any sane way. Here is how [wx does it](https://github.com/wxWidgets/wxWidgets/blob/f4d00c10828c378134ba67de9c473341fa94e174/build/cmake/init.cmake#L11-L52) for comparison.

@Krzmbrzl I wonder if your work on new CMakefiles addresses this already by chance?

Reply:
> which means that C++14 is used if CMAKE_CXX_STANDARD is not set at all.

or it is set to a value < 14.

> What we need to do is to only set the standard to 14 if the compiler doesn't support it in its default configuration.

Why would we want to do such a thing? Projects should always set the C++ standard they use explicitly. Everything else becomes messy pretty quickly.
Given that you are looking at a feature provided as an extension (enabled when using `std=gnu++14` compared to `std=c++14`) the variable you are looking for is `CMAKE_CXX_EXTENSIONS`.

In my cmake PR I have set this variable to `OFF` in order to force the use of portable code that should work on all compilers.

Reply:
> > which means that C++14 is used if CMAKE_CXX_STANDARD is not set at all.
> 
> or it is set to a value < 14.

Yes, but this is as it should be, i.e. not a problem.
 
> > What we need to do is to only set the standard to 14 if the compiler doesn't support it in its default configuration.
> 
> Why would we want to do such a thing? Projects should always set the C++ standard they use explicitly. 

I don't see why would you forbid people from using a later standard, especially if it's the one used by their compiler by default. It also means you can never have any "progressive enhancements", i.e. we're not going to require C++17 just for `[[fallthrough]]` but it's nice to get it when using C++17 compiler anyhow.

I strongly believe that we should never lower the default language version, if anybody really wants to do it, let them set `CMAKE_CXX_STANDARD` to 14 explicitly. This makes more sense by default and is more flexible than the current behaviour. To be honest, I didn't even think that anybody would want to have the current logic, to me it's obviously buggy...

> Everything else becomes messy pretty quickly. Given that you are looking at a feature provided as an extension (enabled when using `std=gnu++14` compared to `std=c++14`) the variable you are looking for is `CMAKE_CXX_EXTENSIONS`.
> 
> In my cmake PR I have set this variable to `OFF` in order to force the use of portable code that should work on all compilers.

The question of `-std=gnuXX` vs `-std=c++XX` is something else and isn't really related to the question of `XX` to use or whether to use any language dialect options at all.

Reply:
If you use the compiler default then you essentially encourage introduction of non-portable code in the sense that the project specifies that it is compatible with a given C++ standard but since someone uses a compiler that defaults to a more recent version they can code up something that won't compile for someone else who in fact has to use the standard the project specifies. 
By being explicit about the used standard version you prevent this as everyone is using the same standard version. 

And with the construct you see in the current cmake files, you ensure that when SOCI is embedded as a subproject, the parent project can use a more recent C++ standard without being held back by the embedded project.

So as you can see we have completely opposite views on this. For me, not having a fixed standard version is an absolute no go.

Reply:
The crucial point is that you can specify the fixed C++ version easily on CMake command line. You can't specify "use my compiler default" easily or, maybe, even at all.

And using the default is much friendlier for non-experts who are much less likely to know how to change this.

Reply:
I disagree. Non-experts should not fiddle with the C++ standard of a given project. 
And I remain confident with the statement that "use whatever standard the compiler happens to default to" is not really good.

Also, I think if you simply don't specify the standard variable at all, you probably end up with the compiler's default.

Reply:
> Also, I think if you simply don't specify the standard variable at all, you probably end up with the compiler's default.

This is what this thread started with: if you don't set anything, SOCI sets the standard to C++14. And I think it shouldn't, if the compiler supports something later by default.

Reply:
I was talking about cmake.

And I think SOCI does the right thing by setting a default standard version. I have not yet encountered any halfway serious project that doesn't specify the used standard version.

Reply:
I'm going to extend the fallthrough definition to include a check for GNU C++ extensions. (The last elif in the following)

```
#if SOCI_CHECK_CXX_STD(201703L)
#    define SOCI_FALLTHROUGH [[fallthrough]]
#elif defined(__has_warning) && SOCI_HAS_CLANG_FEATURE(cxx_attributes)
#    define SOCI_FALLTHROUGH [[clang::fallthrough]]
#elif defined(__GNUC__) && __has_cpp_attribute(fallthrough)
#    define SOCI_FALLTHROUGH [[fallthrough]]
#endif
```

Then it'll also work (as in not complain) in the current scenario.